### PR TITLE
tools: Port unit_tests.py

### LINKED ISSUE #1213: Port tools/unit_tests.py to deno
`tools/unit_tests.py` was written in Python in order to launch subprocesses. (It needs to run the unit tests with various command-line flags.) However, now that Deno has support for running subprocesses via `deno.run()` ( https://deno.land/typedoc/index.html#run ) we can do all of this in JavaScript.

This is relatively easy but it may require some back-and-forth as we try to work out this new API.

### PR DESCRIPTION
**Pull request by DanSnow**
I've port it to deno. Should I also update tools/test.py to execute this script?

Also there is a little difference that I can't not redirect sub process stderr to stdout with current `run` api.

Resolved #1213

### CODE REVIEW

File: tools/unit_tests.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      result.set(buffer.slice(0, nread), tmp.length);
+    } else {
+      result = buffer.slice(0, nread);
+    }
+    if (eof) {
+      break;
+    }
+  }
+  return new TextDecoder("utf-8").decode(result);
+}
```
Comment (ry) [inline tools/unit_tests.ts]: This utility is very useful to have, and I've meant to add it before. I'd rather if we did this in a separate PR tho, and if you put it into `js/io.ts` ... It should use the `deno.Buffer` class, which provides most of the functionality. 

Here's an example: https://github.com/denoland/deno/blob/2ffd78daf9956a24098d1f959f21882e350e9d37/js/fetch.ts#L25-L28

Would you like to do this separate PR for `deno.readAll(r: Reader): Uint8Array` ? 
If so I would ask you look at the implementation in Golang and try to copy it as closely as possible: https://github.com/golang/go/blob/048c9164a0c5572df18325e377473e7893dbfb07/src/io/ioutil/ioutil.go#L16-L46
including this test:
https://github.com/golang/go/blob/048c9164a0c5572df18325e377473e7893dbfb07/src/io/ioutil/example_test.go#L16-L28

File: tools/unit_tests.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+async function runUnitTest2(args: string[]) {
+  const process = run({
+    args,
+    stdin: "null",
+    stdout: "piped"
+  });
+  const { code: errorCode } = await process.status();
+  const output = await readAll(process.stdout);
```
Comment (ry) [inline tools/unit_tests.ts]: Great work! I think we should have a more general purpose utility function which buffers the output. Something like `process.output(): Promise<Uint8Array>`
That should be done in a separate PR too. It can probably be ported from Golang:
https://github.com/golang/go/blob/048c9164a0c5572df18325e377473e7893dbfb07/src/os/exec/exec.go#L498-L520

File: js/run_unit_tests.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+// This test is executed as part of tools/test.py
+// But it can also be run manually: ./target/debug/deno js/run_unit_tests.ts
+// You must have ./tools/http_server.py running in the background if you run
+// this manually.
+
+import { args, exit, run } from "deno";
+
+function extractNumber(pattern: RegExp, str: string) {
+  const matches = pattern.exec(str);
+  if (matches == null || pattern.exec(str) != null) {
```
Comment (piscisaureus) [inline js/run_unit_tests.ts]: ???

File: js/run_unit_tests.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+
+// This test is executed as part of tools/test.py
+// But it can also be run manually: ./target/debug/deno js/run_unit_tests.ts
+// You must have ./tools/http_server.py running in the background if you run
+// this manually.
+
+import { args, exit, run } from "deno";
+
+function extractNumber(pattern: RegExp, str: string) {
```
Comment (piscisaureus) [inline js/run_unit_tests.ts]: This function needs a return type.

File: js/run_unit_tests.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  const matches = pattern.exec(str);
+  if (matches == null || pattern.exec(str) != null) {
+    return;
+  }
+  return Number(matches[1]);
+}
+
+function parseUnitTestOutput(
+  output: string
+): [string | undefined, string | undefined] {
```
Comment (piscisaureus) [inline js/run_unit_tests.ts]: 1. this function should not allow to return undefined for either 'expected' or 'actual'.
2. rather than returning a tuple, return an object `{ actual: number, expected: number }`.

File: js/run_unit_tests.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  const outputBuf = await process.output();
+  const output = new TextDecoder("utf-8").decode(outputBuf);
+
+  const { code: errorCode } = await process.status();
+  if (errorCode !== 0) {
+    console.log(output);
+    exit(errorCode);
+  }
+  const [actual, expected] = parseUnitTestOutput(output);
+  if (actual == null && expected == null) {
```
Comment (piscisaureus) [inline js/run_unit_tests.ts]: I suppose you mean || and not && ?

File: js/run_unit_tests.ts (hunk 3 above)
Comment (DanSnow) [reply js/run_unit_tests.ts]: This make sure that the pattern match exactly once.

**Comment by ry** [comment]
@DanSnow Sorry for not merging this. I'm a bit concerned about changing out test setup right now - in that we may miss errors if we're not careful. I will revisit this patch in a month or so. If you don't mind please leave the branch up. #1234 and #1235 were both very useful improvements.
Closing this PR for now.

**Comment by DanSnow** [comment]
OK. Thanks.
